Extend and improve versioning document (#333)
citynetwork/docs: docs/howto/object-storage/s3/versioning.md

| Delete object                | Object is deleted            | Delete marker is placed, 404 returned |

Once you enable versioning on a bucket, it cannot be turned off - only suspended.
A versioned bucket can be in one of three states: unversioned (default), enabled, or suspended.
When versioning is suspended, new objects are created with a version ID of "null" and will overwrite any existing null version, but all previously versioned objects remain unchanged and accessible.

### Delete markers

When you delete an object in a versioned bucket, the object isn't actually removed.
Instead, object service creates a special placeholder called a delete marker that acts as the current version of the object.
This delete marker has its own unique version ID and causes object service to return a 404 "Not found" error when you try to retrieve the object normally.
However, all previous versions remain intact and accessible if you specify their version IDs.
Delete markers themselves can be deleted - when you delete a delete marker, the previous version becomes current again, effectively "undeleting" the object.

> To permanently remove an object version, you must delete it by specifying its version ID.

Author: Was trying to follow convention [here](https://docs.cleura.cloud/howto/object-storage/s3/object-lock/) more, but of course I can move into backgroud.

* If the object does not already exist, it is created (as in an
  unversioned bucket).

> * If the object does already exist, it is not replaced. Instead, a new version becomes

Reviewer: ```suggestion
* If the object already exists, it is not replaced.
```

> Instead, the new version becomes the current one.

Reviewer 2: I've applied this, but this is actually wrong as the two-space indent is missing before "Instead". I'll fix this myself.

>   the current version in addition to the old one.

Reviewer: ```suggestion
Instead, the new version becomes the current one.
```

* If the object does not already exist, it is created (as in an
  unversioned bucket).
* If the object does already exist, it is not replaced. Instead, a new version becomes

>   the current version in addition to the old one.

Reviewer 2: Please remove this change; it is wrong. "A new version becomes the current version in addition to the old one" means that there are subsequently two "current" versions, which is not true.

Author: Previous "a new version appears in addition to the old one." doesn't explain which one will be curent default, so which one will receive "Is Latest: True" param. My will is to express it in a way that always new overwrite will become this latest, so current one.

specifying a version identifier, your S3 client will download the
latest version of that object.

## Deleting a versioned object

Like the commands to *create* objects, the commands to *delete* them
behave differently once object versioning is enabled on a bucket.

The command to delete an object will not delete it, but
put a delete marker on it instead. This will keep all the versions

> but replay "Not found" 404 on a request without specified valid version id.

Reviewer: ```suggestion
but return a "Not found" 404 on any request not specifying a valid version id.
```

specifying a version identifier, your S3 client will download the
latest version of that object.

## Deleting a versioned object

Like the commands to *create* objects, the commands to *delete* them
behave differently once object versioning is enabled on a bucket.

The command to delete an object will not delete it, but
put a delete marker on it instead. This will keep all the versions

> but replay "Not found" 404 on a request without specified valid version id.

Reviewer 2: I don't understand, and I don't think this is factually correct. Specifically, I don't think it's correct to delete the first sentence in this paragraph. What is your source for this?

Author: I think that stated here "but revert it to the prior version" is actually incorrect. Can you show me where there is an error in my statement ?

Output of a test on an object service:

```
1. On a versioned bucket I do create multiple versions of a object

▶ Creating multiple versions of 'test-overwrite.txt'...
✓   Created version 2 - ID: GY5YGM6qkT8OdusRdBlaul1c.DoL2m3
✓   Created version 3 - ID: Apn-Tw.wWKUVs6TM9PxUjKEZgzQ201e
✓   Created version 4 - ID: cEFKyOZOa8YeNXMYFdYNn-SSIlIKkIb

▶ Listing all versions of 'test-overwrite.txt'...
ℹ
Found 4 versions:
  1. Version ID: cEFKyOZOa8YeNXMYFdYNn-SSIlIKkIb
     Size: 57 bytes
     Last Modified: 2025-05-23 14:19:44.400000+00:00
     ETag: "f3b88c42e794440a9c4f9920f86cd005"
     Is Latest: True
  2. Version ID: Apn-Tw.wWKUVs6TM9PxUjKEZgzQ201e
     Size: 57 bytes
     Last Modified: 2025-05-23 14:19:43.082000+00:00
     ETag: "f50e24ba03b69df9c0cd9dca23869279"
     Is Latest: False
  3. Version ID: GY5YGM6qkT8OdusRdBlaul1c.DoL2m3
     Size: 57 bytes
     Last Modified: 2025-05-23 14:19:41.767000+00:00
     ETag: "92c637f74d0bfd7b204f46dd0439ba25"
     Is Latest: False
  4. Version ID: .AyXiqOYEthH5HFt-pSBEHSTKLD0YR9
     Size: 15 bytes
     Last Modified: 2025-05-23 14:19:41.445000+00:00
     ETag: "fcf830e17b79d01997f558a0478e77de"
     Is Latest: False
     
2. Then, I delete the object with s3.delete_object(Bucket=bucket_name, Key=key), then try to acess with s3.get_object(Bucket=bucket_name, Key=key) /404 is expected here not a reverted version!/ and try to access deleted (marked) version with version scoped query s3.get_object(Bucket=bucket_name,Key=key,VersionId=latest_version_id).

▶ Deleting object (creating delete marker)...
✓ Delete marker created - Version ID: SqIVcufE89OvPNFuUmrqgYEo5JySGuY
ℹ Delete marker: True

▶ Attempting to access deleted object (expecting 404)...
✓ Got expected 404 (NoSuchKey) error!
ℹ Error message: None

▶ Accessing previous version directly (Version ID: cEFKyOZOa8YeNXMYFdYNn-SSIlIKkIb)...
✓ Previous version still accessible!
ℹ Content preview: Version 4 content - Updated at 2025-05-23T16:19:44...

3. If we will list object then we can see

▶ Listing all objects and versions in bucket...
ℹ
Regular versions found: 4
  - Key: test-overwrite.txt
    Version ID: cEFKyOZOa8YeNXMYFdYNn-SSIlIKkIb
    Size: 57 bytes
    Is Latest: False
  - Key: test-overwrite.txt
    Version ID: Apn-Tw.wWKUVs6TM9PxUjKEZgzQ201e
    Size: 57 bytes
    Is Latest: False
  - Key: test-overwrite.txt
    Version ID: GY5YGM6qkT8OdusRdBlaul1c.DoL2m3
    Size: 57 bytes
    Is Latest: False
  - Key: test-overwrite.txt
    Version ID: .AyXiqOYEthH5HFt-pSBEHSTKLD0YR9
    Size: 15 bytes
    Is Latest: False
ℹ
Delete markers found: 1
  - Key: test-overwrite.txt
    Version ID: SqIVcufE89OvPNFuUmrqgYEo5JySGuY
    Is Latest: True
```
so no "is latest: true" on any version but it's set on delete marker instead. So there is no revert at any point, this would be ultimately unintuitive as well.
